scipy/sparse/csgraph/_flow.pyx

```diff
@@ -407,3 +431,207 @@ cdef ITYPE_t[:] _edmonds_karp(
             # If no augmenting path could be found, we're done.
             break
     return flow
+
+cdef bint _build_level_graph(
+        const ITYPE_t[:] edge_ptr, # IN
+        const ITYPE_t source, # IN
+        const ITYPE_t sink, # IN
+        const ITYPE_t[:] capacities, # IN
+        const ITYPE_t[:] heads, # IN
+        ITYPE_t[:] levels, # IN/OUT
+        ITYPE_t[:] q, # IN/OUT
+    ) nogil:
+    """Builds layered graph from input graph using breadth first search.
+
+    Parameters
+    ----------
+    edge_ptr : memoryview of length :math:`|V| + 1`
+        For a given vertex ``v``, the edges whose tail is ``v`` are those between
+        ``edge_ptr[v]`` and ``edge_ptr[v + 1] - 1``.
+    source : int
+        The source vertex.
+    sink : int
+        The sink vertex.
+    capacities : memoryview of length :math:`|E|`
+        For a given edge ``e``, ``capacities[e]`` is the capacity of ``e``.
+    heads : memoryview of length :math:`|E|`
+        For a given edge ``e``, ``heads[e]`` is the head vertex of ``e``.
+    levels: memoryview of length :math:`|E|`
+        For a given vertex ``v``, ``levels[v]`` is the level of ``v`` in 
+        the layered graph of input graph.
+    q : memoryview of length :math:`|E|`
+        Queue to be used in breadth first search. Passed to avoid repeated 
+        queue creation inside this function.
+
+    Returns
+    -------
+    bool:
+        ``True`` if the layered graph creation was successful,
+        otherwise ``False``.
+
+    """
+    cdef ITYPE_t n_verts = edge_ptr.shape[0] - 1
+
+    cdef ITYPE_t cur, start, end, dst_vertex, e
+
+    q[0] = source
+    start = 0
+    end = 1
+    levels[source] = 0
+
+    while start != end:
+        cur = q[start]
+        start += 1
+        if start == n_verts:
```

Comment:
You only visit each vertex once, right? If so, you'll never loop around in this circular queue, and this check is redundant (similarly with `end == n_verts` below). (Same actually for with the Edmonds--Karp implementation, but that's unrelated to the PR.)

Reply:
Addressed.